### PR TITLE
Brcampbe/windows dcos

### LINKED ISSUE #1078: Add osType=Windows for DC/OS deployment
**Is this a request for help?**:
No, other than we will want help to release it once the work is complete.
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
Feature request that we are adding
---

**What version of acs-engine?**:
v.next
---

We are adding the osType=Windows for DC/OS deployment. We are doing it based on Kubernetes and Swarm and based on some prototype we have done to generate a Windows VM running a DC/OS engine.

In both cases, be ready for followup questions, and please respond in a timely
manner.  If we can't reproduce a bug or think a feature already exists, we
might close your issue.  If we're wrong, PLEASE feel free to reopen it and
explain why.
**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
DC/OS

**What happened**:

**What you expected to happen**:

**How to reproduce it** (as minimally and precisely as possible):

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by yakman2020**
**What this PR does / why we need it**: Support for windows agents for DCOS 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #1078

**Special notes for your reviewer**: Most changes in the form of new files.  

**Release note**:

```release-note
```

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/1132)

### CODE REVIEW

**Comment by seanknox** [comment]
@acs-bot test this please

File: Makefile [hunk 1]
```diff
@@ -57,6 +57,8 @@ checksum:
 .PHONY: clean
 clean:
 	@rm -rf $(BINDIR) ./_dist
+	go build -v -gcflags="-N -l" -ldflags="-X github.com/Azure/acs-engine/cmd.BuildSHA=${VERSION} -X github.com/Azure/acs-engine/cmd.BuildTime=${BUILD}"
+	cd test/acs-engine-test; go build -v
```
Comment (seanknox) [inline Makefile]: `make clean` should only do that—this logic is already handled in `make build`.

File: parts/dcosWindowsAgentResourcesVmas.t.orig [hunk 2]
```diff
@@ -0,0 +1,240 @@
+    {
+      "apiVersion": "[variables('apiVersionDefault')]",
```
Comment (seanknox) [inline parts/dcosWindowsAgentResourcesVmas.t.orig]: This file is an artifact from a merge conflict, you can delete it.

File: parts/dcosWindowsAgentResourcesVmss.t.orig [hunk 3]
```diff
@@ -0,0 +1,204 @@
+    {
+      "apiVersion": "[variables('apiVersionDefault')]",
```
Comment (seanknox) [inline parts/dcosWindowsAgentResourcesVmss.t.orig]: You can delete this file too.

File: parts/dcoscustomdata190.t [hunk 4]
```diff
@@ -1,6 +1,5 @@
 bootcmd:
-- bash -c "if [ ! -f /var/lib/sdb-gpt ];then echo DCOS-5890;parted -s /dev/sdb mklabel
-  gpt;touch /var/lib/sdb-gpt;fi"
+- bash -c "if [ ! -f /var/lib/sdb-gpt ];then echo DCOS-5890;parted -s /dev/sdb mklabel gpt;touch /var/lib/sdb-gpt;fi"
```
Comment (seanknox) [inline parts/dcoscustomdata190.t]: This is a little harder to grok than the two line version.

File: pkg/acsengine/templates.go [hunk 5]
```diff
@@ -0,0 +1,1707 @@
+// Code generated by go-bindata.
```
Comment (seanknox) [inline pkg/acsengine/templates.go]: You might be working off of an old branch as `pkg/acsengine/templates.go` is not checked into source control. Please delete it.

File: pkg/api/apiloader.go [hunk 6]
```diff
@@ -9,6 +9,7 @@ import (
 	"github.com/Azure/acs-engine/pkg/api/v20160930"
 	"github.com/Azure/acs-engine/pkg/api/v20170131"
 	"github.com/Azure/acs-engine/pkg/api/v20170701"
+	"github.com/Azure/acs-engine/pkg/api/v20170801"
```
Comment (seanknox) [inline pkg/api/apiloader.go]: @anhowe Should DCOS Windows support be in vlabs only for now?

**Comment by yakman2020** [comment]
@acs-bot test this please

File: parts/dcoscustomdata190.t (hunk 4 above)
Comment (yakman2020) [inline parts/dcoscustomdata190.t]: done as requested.

**Comment by seanknox** [comment]
@acs-bot test this please

**Comment by seanknox** [comment]
@acs-bot test this please

**Comment by seanknox** [comment]
@acs-bot test this please

**Comment by seanknox** [comment]
@dmitsh believe you're on the docket to help @yakman2020 on this one...CI doesn't seem to be running for this branch, or at least the bot isn't kicking off jobs.

File: examples/windows/dcos-hybrid.json.env [hunk 7]
```diff
@@ -0,0 +1 @@
+export EXPECTED_NODE_COUNT=2
```
Comment (dmitsh) [inline examples/windows/dcos-hybrid.json.env]: Why do you need this?
By default we calculate the node count by parsing api model file.

File: pkg/acsengine/engine.go [hunk 8]
```diff
@@ -642,12 +651,17 @@ func (t *TemplateGenerator) getTemplateFuncMap(cs *api.ContainerService) map[str
 			return fmt.Sprintf("\"customData\": \"[base64(concat('#cloud-config\\n\\n', '%s'))]\",", str)
 		},
 		"GetDCOSAgentCustomData": func(profile *api.AgentPoolProfile) string {
-			agentProvisionScript := getDCOSAgentProvisionScript(profile)
+		    agentProvisionScript := getDCOSAgentProvisionScript(profile)
 			attributeContents := getDCOSAgentCustomNodeLabels(profile)
+log.Info("GetDCOSAgentCustomData2\n");
```
Comment (dmitsh) [inline pkg/acsengine/engine.go]: not very descriptive log message

File: examples/windows/dcos-hybrid.json.env (hunk 7 above)
Comment (yakman2020) [inline examples/windows/dcos-hybrid.json.env]: removed the file.

File: pkg/acsengine/engine.go (hunk 8 above)
Comment (yakman2020) [inline pkg/acsengine/engine.go]: removed the log message

File: parts/dcosmastervars.t [hunk 9]
```diff
@@ -36,8 +46,15 @@
     "masterNSGName": "[concat(variables('orchestratorName'), '-master-nsg-', variables('nameSuffix'))]",
     "masterPublicIPAddressName": "[concat(variables('orchestratorName'), '-master-ip-', variables('masterEndpointDNSNamePrefix'), '-', variables('nameSuffix'))]",
     "apiVersionStorage": "2015-06-15",
-    "storageAccountBaseName": "[uniqueString(concat(variables('masterEndpointDNSNamePrefix'),variables('location'),variables('orchestratorName')))]",
-    "masterStorageAccountExhibitorName": "[concat(variables('storageAccountBaseName'), 'exhb0')]",
+{{if .HasWindows}}
```
Comment (jackfrancis) [inline parts/dcosmastervars.t]: These additional windows-specific vars should go inside the existing `{{if .HasWindows}}` block above. (see line 26)

File: parts/dcosagentvars.t [hunk 10]
```diff
@@ -31,4 +36,12 @@
     "{{.Name}}LbIPConfigID": "[concat(variables('{{.Name}}LbID'),'/frontendIPConfigurations/', variables('{{.Name}}LbIPConfigName'))]", 
     "{{.Name}}LbIPConfigName": "[concat(variables('orchestratorName'), '-{{.Name}}-', variables('nameSuffix'))]", 
     "{{.Name}}LbName": "[concat(variables('orchestratorName'), '-{{.Name}}-', variables('nameSuffix'))]",
-{{end}}
\ No newline at end of file
+     {{if .IsWindows}}
+        "{{.Name}}WindowsRDPNatRangeStart": 3389,
+        "{{.Name}}WindowsRDPEndRangeStop": "[add(variables('{{.Name}}WindowsRDPNatRangeStart'), add(variables('{{.Name}}Count'),variables('{{.Name}}Count')))]",
+    {{end}}
+    "{{.Name}}windowsAgentCustomScriptArguments": "[concat('$arguments = ', variables('singleQuote'),'-subnet ', variables('{{.Name}}Subnet'), ' -MasterCount ', variables('masterCount'), ' -firstMasterIP ', parameters('firstConsecutiveStaticIP'), ' -bootstrapUri ', '\"', variables('dcosWindowsBootstrapURL'), '\"', ' -isAgent $true -isPublic:$true',  variables('singleQuote'),' ; ')]",
```
Comment (jackfrancis) [inline parts/dcosagentvars.t]: Shouldn't this be a windows-only change? (i.e., inside the `{{if .IsWindows}}` branch)?

File: parts/dcosagentvars.t [hunk 11]
```diff
@@ -31,4 +36,12 @@
...
+{{else}}
+    "{{.Name}}windowsAgentCustomScriptArguments": "[concat('$arguments = ', variables('singleQuote'),'-subnet ', variables('{{.Name}}Subnet'), ' -MasterCount ', variables('masterCount'), ' -firstMasterIP ', parameters('firstConsecutiveStaticIP'), ' -bootstrapUri ', '\"', variables('dcosWindowsBootstrapURL'), '\"', ' -isAgent $true',  variables('singleQuote'),' ; ')]",
```
Comment (jackfrancis) [inline parts/dcosagentvars.t]: ditto

File: parts/dcosagentvars.t [hunk 12]
```diff
@@ -31,4 +36,12 @@
...
+{{end}}
+    "{{.Name}}windowsAgentCustomScript": "[concat('powershell.exe -ExecutionPolicy Unrestricted -command \"', variables('{{.Name}}windowsAgentCustomScriptArguments'), variables('windowsCustomScriptSuffix'), '\" > %SYSTEMDRIVE%\\AzureData\\dcosWindowsProvision.log 2>&1')]",
```
Comment (jackfrancis) [inline parts/dcosagentvars.t]: ditto

**Comment by yakman2020** [comment]
Jack's requested changes added

**Comment by jackfrancis** [comment]
Thanks for hanging with us, LGTM!

**Pull request merged**